internal/safebuffer/safereader/safereader.go

```diff
@@ -0,0 +1,160 @@
+package safereader
+
+import (
+	"bytes"
+	"io"
+	"log/slog"
+	"sync"
+)
+
+// Reader is a goroutine safe io.ReadCloser that makes initial data available to
+// callers as well as new data that arrives later via channel.
+type Reader struct {
+	mu        sync.Mutex
+	buf       *bytes.Buffer
+	close     func()
+	closed    <-chan struct{}
+	dataCh    <-chan []byte
+	dataAvail chan struct{}
+	closeOnce sync.Once
+}
+
+var _ io.ReadCloser = (*Reader)(nil)
+
+// jobIsRunning doesn't have a race because Buffer.NewReader calls it while
+// holding a lock that prevents new data, or the close of jobDone, from being
+// processed.
+func jobIsRunning(jobDone <-chan struct{}) bool {
+	select {
+	case <-jobDone:
+		return false
+	default:
+		return true
+	}
+}
+
+type Channels struct {
+	Data   chan []byte
+	Closed <-chan struct{}
+}
+
+// New returns a new Reader that keeps a copy of data and will append any data
+// that arrives on the returned Data channel. If the job is no longer running
+// when created, it is treated as a closed reader and no new data can be added.
+// Calls to Close() in this case are ignored.
+func New(data []byte, jobDone <-chan struct{}) (*Reader, *Channels) {
+	closed := make(chan struct{})
+	closeFn := func() { close(closed) }
+
+	buf := make([]byte, len(data))
+	copy(buf, data)
+
+	r := Reader{
+		close:  closeFn,
+		closed: closed,
+		buf:    bytes.NewBuffer(buf),
+	}
+
+	var channels *Channels
+
+	if jobIsRunning(jobDone) {
+		dataCh := make(chan []byte)
+		channels = &Channels{
+			Data:   dataCh,
+			Closed: closed,
+		}
+		r.dataCh = dataCh
+		r.dataAvail = make(chan struct{})
+
+		go r.receiver()
+	} else {
+		r.Close()
```

Comment:
why closing the reader here?

Reply:
Readers are closed when jobs complete (or when clients call `Close()`), but in this case the job has already completed and we know no new job output will be coming in. As a result, the state of the reader is simplified in that it only needs to be a buffer for the data it has when it is created. It doesn't need to be added to the map in the `SafeBuffer`, it doesn't need a goroutine, the calls to `Read()` can just defer to the internal buffer. By closing it immediately, ensure that clients won't block indefinitely for jobs that have already completed and for which they haven't called `Close()` in a separate goroutine. I'll add a comment to clarify.